// iamport-react-native/lib/commonjs/components/Loading/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const logo = require('../../img/iamport-logo.png');
function Loading() {
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.container
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
    source: logo
  }), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.text
  }, "\uC7A0\uC2DC\uB9CC \uAE30\uB2E4\uB824\uC8FC\uC138\uC694..."));
}
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    flexDirection: 'column',
    justifyContent: 'center'
  },
  text: {
    fontSize: 20,
    marginTop: 20,
    lineHeight: 25
  }
});
var _default = Loading;
exports.default = _default;
