// @ckeditor/ckeditor5-table/src/index.js
/**
 * @license Copyright (c) 2003-2024, CKSource Holding sp. z o.o. All rights reserved.
 * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
 */
/**
 * @module table
 */
export { default as PlainTableOutput } from './plaintableoutput.js';
export { default as Table } from './table.js';
export { default as TableEditing } from './tableediting.js';
export { default as TableUI } from './tableui.js';
export { default as TableToolbar } from './tabletoolbar.js';
export { default as TableCellProperties } from './tablecellproperties.js';
export { default as TableCellPropertiesEditing } from './tablecellproperties/tablecellpropertiesediting.js';
export { default as TableCellPropertiesUI } from './tablecellproperties/tablecellpropertiesui.js';
export { default as TableCellWidthEditing } from './tablecellwidth/tablecellwidthediting.js';
export { default as TableProperties } from './tableproperties.js';
export { default as TablePropertiesEditing } from './tableproperties/tablepropertiesediting.js';
export { default as TablePropertiesUI } from './tableproperties/tablepropertiesui.js';
export { default as TableCaption } from './tablecaption.js';
export { default as TableCaptionEditing } from './tablecaption/tablecaptionediting.js';
export { default as TableCaptionUI } from './tablecaption/tablecaptionui.js';
export { default as TableClipboard } from './tableclipboard.js';
export { default as TableMouse } from './tablemouse.js';
export { default as TableKeyboard } from './tablekeyboard.js';
export { default as TableSelection } from './tableselection.js';
export { default as TableUtils } from './tableutils.js';
export { default as TableColumnResize } from './tablecolumnresize.js';
export { default as TableColumnResizeEditing } from './tablecolumnresize/tablecolumnresizeediting.js';
import './augmentation.js';
