// iamport-react-native/lib/commonjs/components/ErrorOnParams/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const logo = require('../../img/iamport-logo.png');
function ErrorOnParams(_ref) {
  let {
    message
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.container
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
    source: logo
  }), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.text
  }, message));
}
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center'
  },
  text: {
    textAlign: 'center',
    fontSize: 16,
    marginTop: 20,
    lineHeight: 25
  },
  button: {
    backgroundColor: '#fff',
    color: '#ff0000',
    borderColor: '#333',
    borderWidth: 1,
    borderRadius: 3,
    paddingTop: 10,
    paddingBottom: 10,
    paddingLeft: 20,
    paddingRight: 20,
    fontSize: 14,
    marginTop: 20,
    fontWeight: 'bold'
  }
});
var _default = ErrorOnParams;
exports.default = _default;
